django_cloud_tasks/models.py

```diff
@@ -0,0 +1,118 @@
+from typing import Optional
+from django.db import transaction
+from django.apps import apps
+
+from datetime import datetime
+from django.db import models
+from django_cloud_tasks import tasks
+
+class Pipeline(models.Model):
+    name = models.CharField(max_length=50)
+
+    def start(self):
+        routines = self.routines.filter(required_routine_vertices__id__isnull=True)
+        for routine in routines:
+            routine.enqueue()
+
+    def revert(self):
+        routines = self.routines.filter(next_routine_vertices__id__isnull=True)
+        for routine in routines:
+            routine.revert()
+
+class Routine(models.Model):
+    class Statuses(models.TextChoices):
+        PENDING = ("pending", "Pending")
+        SCHEDULED = ("scheduled", "scheduled")
+        RUNNING = ("running", "running")
+        COMPLETED = ("completed", "Completed")
+        FAILED = ("failed", "Failed")
+        REVERTING = ("reverting", "Reverting")
+        REVERTED = ("reverted", "Reverted")
+        ABORTED = ("aborted", "Aborted")
+
+    task_name = models.CharField(max_length=50)
+    pipeline = models.ForeignKey(
+        to="django_cloud_tasks.Pipeline",
+        related_name="routines",
+        on_delete=models.PROTECT,
+    )
+    body = models.JSONField(default=dict)
+    attempt_count = models.PositiveIntegerField(default=0)
+    retries_exhausted_limit = models.PositiveIntegerField(null=True)
+    output = models.JSONField(null=True, blank=True)
+    starts_at = models.DateTimeField(null=True, blank=True)
+    ends_at = models.DateTimeField(null=True, blank=True)
+    status = models.CharField(
+        max_length=20,
+        choices=Statuses.choices,
+        default=Statuses.PENDING,
+    )
+    created_at = models.DateTimeField(
+        auto_now_add=True,
+    )
+    updated_at = models.DateTimeField(
+        auto_now=True,
+    )
+    next_routines = models.ManyToManyField(
+        to='Routine',
+        through='RoutineVertice',
+        through_fields=('routine', 'next_routine'),
+        related_name="dependencies"
```

Comment:
The names of both sides are weird:
`routine.next_routines.all()` and `routine.dependencies.all()`

Both should have [or not] the "dependencies" prefix.